crates/runtime/src/logic.rs

```diff
@@ -627,81 +659,238 @@ impl VMHostFunctions<'_> {
         Ok(())
     }
 
-    // ========== SIMPLIFIED BLOB HOST FUNCTIONS ==========
+    // ========== CHUNKED BLOB HOST FUNCTIONS ==========
 
-    /// Store blob data and return the blob ID in a register
-    /// Returns: 1 if successful, 0 if failed
-    pub fn store_blob(
-        &mut self,
-        data_ptr: u64,
-        data_len: u64,
-        register_id: u64,
-    ) -> VMLogicResult<u32> {
-        // Check if blob functionality is available
-        let node_client = match &self.borrow_logic().node_client {
-            Some(client) => client.clone(),
-            None => return Err(VMLogicError::HostError(HostError::BlobsNotSupported)),
-        };
+    /// Create a new blob for writing
+    /// Returns: file descriptor (u64) for writing operations
+    pub fn blob_create(&mut self) -> VMLogicResult<u64> {
+        if self.borrow_logic().node_client.is_none() {
+            return Err(VMLogicError::HostError(HostError::BlobsNotSupported));
+        }
+
+        if self.borrow_logic().blob_handles.len()
+            >= self.borrow_logic().limits.max_blob_handles as usize
+        {
+            return Err(VMLogicError::HostError(HostError::TooManyBlobHandles {
+                max: self.borrow_logic().limits.max_blob_handles,
+            }));
+        }
+
+        let fd = self.with_logic_mut(|logic| {
+            let fd = logic.next_blob_fd;
+            logic.next_blob_fd += 1;
+
+            let (data_sender, data_receiver) = mpsc::unbounded_channel();
+
+            let node_client = logic.node_client.clone().unwrap();
+
+            let completion_handle = tokio::spawn(async move {
+                let stream = UnboundedReceiverStream::new(data_receiver);
+
+                let byte_stream =
+                    stream.map(|data: Vec<u8>| Ok::<bytes::Bytes, std::io::Error>(data.into()));
+                let reader = byte_stream.into_async_read();
+
+                node_client.add_blob(reader, None, None).await
+            });
+
+            let handle = BlobHandle::Write(BlobWriteHandle {
+                sender: data_sender,
+                completion_handle,
+            });
```

Comment:
all this can happen outside the closure no? we should only need to call `with_logic_mut` specifically to increment the fd and insert the handle, helps readability